mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,247 @@
+import os
+import tempfile
+from time import sleep
+from typing import ClassVar, Optional
+
+from kubernetes import client, config, utils, watch
+
+from mlem.config import project_config
+from mlem.core.objects import (
+    DeployState,
+    DeployStatus,
+    MlemBuilder,
+    MlemDeployment,
+    MlemEnv,
+    MlemModel,
+)
+from mlem.runtime.client import Client, HTTPClient
+from mlem.runtime.server import Server
+from mlem.ui import EMOJI_BUILD, EMOJI_FAIL, EMOJI_OK, echo, set_offset
+
+from ..docker.base import (
+    DockerDaemon,
+    DockerEnv,
+    DockerImage,
+    DockerRegistry,
+    generate_docker_container_name,
+)
+from ..docker.helpers import build_model_image
+from .context import K8sYamlBuildArgs, K8sYamlGenerator
+
+POD_STATE_MAPPING = {
+    "Pending": DeployStatus.STARTING,
+    "Running": DeployStatus.RUNNING,
+    "Succeeded": DeployStatus.STOPPED,
+    "Failed": DeployStatus.CRASHED,
+    "Unknown": DeployStatus.UNKNOWN,
+}
+
+
+class K8sDeploymentState(DeployState):
+    type: ClassVar = "kubernetes"
+    image: Optional[DockerImage] = None
+    deployment_name: Optional[str] = None
+
+
+class K8sDeployment(MlemDeployment, K8sYamlBuildArgs):
+    type: ClassVar = "kubernetes"
+    server: Optional[Server] = None
+    state_type: ClassVar = K8sDeploymentState
+    registry: Optional[DockerRegistry] = DockerRegistry()
+    daemon: Optional[DockerDaemon] = DockerDaemon(host="")
+    kube_config_file_path: Optional[str] = None
+
+    def load_kube_config(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+    def find_index(self, nodes_list, node_name):
+        for i, each_node in enumerate(nodes_list):
+            if each_node.metadata.name == node_name:
+                return i
+        return -1
+
+    def _get_client(self, state: K8sDeploymentState) -> Client:
+        self.load_kube_config()
+        service = client.CoreV1Api().list_namespaced_service(
+            f"mlem-{state.deployment_name}-app"
+        )
+        if self.service_type == "NodePort":
```

Comment:
this is good opportunity to make `service_type` an `MlemABC`, esp if we plan to add others later

Reply:
There are only 3 types of services -- `ClusterIP`, `NodePort` and `LoadBalancer`
Out of these 3, only the last 2 are reachable from outside the Kubernetes cluster. Hence only 2 if/else.
But I have added handling of invalid values supplied.